extra_geom/detectors.py

```diff
@@ -768,6 +768,81 @@ def offset(self, shift, *, modules=np.s_[:], tiles=np.s_[:]):
         ])
 
 
+class CrystFEL_Geometry(DetectorGeometryBase):
+    """A generic detector layout based on the CrystFEL geom file.
+
+    The coordinates used in this class are 3D (x, y, z), and represent metres.
+    """
+    detector_type_name = 'Generic Detector'
+    pixel_size = None   # should be in meters
+    frag_ss_pixels = None
+    frag_fs_pixels = None
+    expected_data_shape = ()
+    n_quads = 0
+    n_modules = None
+    n_tiles_per_module = None
+
+    def __init__(self, pixel_size: float, fast_pixels: int, slow_pixels: int,
+                 corner_coordinates: [np.ndarray] = [np.zeros(3)],
+                 fs_vec: np.ndarray = np.array([0, 1, 0]),
+                 ss_vec: np.ndarray = np.array([1, 0, 0]),
+                 n_tiles_per_module: int = 1,
+                 tile_offset: float = None,
+                 tile_vec: np.ndarray = None,
+                 ):
+        """ Create a generic detector:
+
+        `pixel_size`: the size of a pixel in meters (reversed CrystFEL's `res`)
+        `fast_pixels`, `slow_pixels`: the size of a tile along fast and slow axes
+        `corner_coordinates`: 3D coordinates of the first pixel of each module
+        `n_modules`: the number of modules, default=1
+        `n_tiles_per_module`:  the number of tiles in each module, default=1
+        `tile_offset`: the gap between two tiles, default=pixel_size
+        `tile_vec`: the direction of tile replication, default=[1, 0, 0]
+
+        """
+        self.pixel_size = pixel_size
+        self.frag_fs_pixels = fast_pixels
+        self.frag_ss_pixels = slow_pixels
+        self.n_modules = len(corner_coordinates)
+        self.n_tiles_per_module = n_tiles_per_module
+        self.expected_data_shape = (self.n_modules,
+                                    self.n_tiles_per_module * self.frag_ss_pixels,
```

Comment:
You're assuming that the tiles in a module are always arranged in one row along the slow-scan axis. Of the specific detectors we support, this is true for AGIPD, but not for LPD, DSSC or JUNGFRAU. I think we'll want to provide a bit more flexibility here. What about something like this:

```python
GenericDetector(
    ...
    pixels_in_tile = (64, 128),  # slow-scan, fast-scan
    tiles_in_module = (8, 1),    # slow-scan, fast-scan
)

Reply:
Oh, no, maybe that's not true for LPD...

Reply:
I assume that tiles are always arranged in one row along the `tile_vec` vector. I agree that this is a restriction. But with three different vectors we have (`fs_vec`, `ss_vec`, `tile_vec`) one may define rather rich spectrum of detectors, see the documentation I prepared. Allowing tiles to be arranged in 2D will immediately make things much more complex. I assume those types of detectors (LPD-like) may be described via a `geom`-file and loaded with `from_crystfel_geom` function, which I will introduce later.

Reply:
@takluyver Can we agree on the current implementation? (See the docs.)